File: src/main/java/io/github/erp/service/FxTransactionChannelTypeService.java
package io.github.erp.service;

/*-
 * Erp System - Mark VI No 3 (Phoebe Series) Server ver 1.5.4
 * Copyright © 2021 - 2023 Edwin Njeru (mailnjeru@gmail.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import io.github.erp.service.dto.FxTransactionChannelTypeDTO;
import java.util.Optional;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Service Interface for managing {@link io.github.erp.domain.FxTransactionChannelType}.
 */
public interface FxTransactionChannelTypeService {
    /**
     * Save a fxTransactionChannelType.
     *
     * @param fxTransactionChannelTypeDTO the entity to save.
     * @return the persisted entity.
     */
    FxTransactionChannelTypeDTO save(FxTransactionChannelTypeDTO fxTransactionChannelTypeDTO);

    /**
     * Partially updates a fxTransactionChannelType.
     *
     * @param fxTransactionChannelTypeDTO the entity to update partially.
     * @return the persisted entity.
     */
    Optional<FxTransactionChannelTypeDTO> partialUpdate(FxTransactionChannelTypeDTO fxTransactionChannelTypeDTO);

    /**
     * Get all the fxTransactionChannelTypes.
     *
     * @param pageable the pagination information.
     * @return the list of entities.
     */
    Page<FxTransactionChannelTypeDTO> findAll(Pageable pageable);

    /**
     * Get the "id" fxTransactionChannelType.
     *
     * @param id the id of the entity.
     * @return the entity.
     */
    Optional<FxTransactionChannelTypeDTO> findOne(Long id);

    /**
     * Delete the "id" fxTransactionChannelType.
     *
     * @param id the id of the entity.
     */
    void delete(Long id);

    /**
     * Search for the fxTransactionChannelType corresponding to the query.
     *
     * @param query the query of the search.
     *
     * @param pageable the pagination information.
     * @return the list of entities.
     */
    Page<FxTransactionChannelTypeDTO> search(String query, Pageable pageable);
}

File: erp-client/src/main/webapp/app/entities/gdi/fx-transaction-channel-type/service/fx-transaction-channel-type.service.ts
///
/// Erp System - Mark X No 10 (Jehoiada Series) Client 1.7.8
/// Copyright © 2021 - 2024 Edwin Njeru (mailnjeru@gmail.com)
///
/// This program is free software: you can redistribute it and/or modify
/// it under the terms of the GNU General Public License as published by
/// the Free Software Foundation, either version 3 of the License, or
/// (at your option) any later version.
///
/// This program is distributed in the hope that it will be useful,
/// but WITHOUT ANY WARRANTY; without even the implied warranty of
/// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
/// GNU General Public License for more details.
///
/// You should have received a copy of the GNU General Public License
/// along with this program. If not, see <http://www.gnu.org/licenses/>.
///

import { Injectable } from '@angular/core';
import { HttpClient, HttpResponse } from '@angular/common/http';
import { Observable } from 'rxjs';

import { isPresent } from 'app/core/util/operators';
import { ApplicationConfigService } from 'app/core/config/application-config.service';
import { createRequestOption } from 'app/core/request/request-util';
import { SearchWithPagination } from 'app/core/request/request.model';
import { IFxTransactionChannelType, getFxTransactionChannelTypeIdentifier } from '../fx-transaction-channel-type.model';

export type EntityResponseType = HttpResponse<IFxTransactionChannelType>;
export type EntityArrayResponseType = HttpResponse<IFxTransactionChannelType[]>;

@Injectable({ providedIn: 'root' })
export class FxTransactionChannelTypeService {
  protected resourceUrl = this.applicationConfigService.getEndpointFor('api/fx-transaction-channel-types');
  protected resourceSearchUrl = this.applicationConfigService.getEndpointFor('api/_search/fx-transaction-channel-types');

  constructor(protected http: HttpClient, protected applicationConfigService: ApplicationConfigService) {}

  create(fxTransactionChannelType: IFxTransactionChannelType): Observable<EntityResponseType> {
    return this.http.post<IFxTransactionChannelType>(this.resourceUrl, fxTransactionChannelType, { observe: 'response' });
  }

  update(fxTransactionChannelType: IFxTransactionChannelType): Observable<EntityResponseType> {
    return this.http.put<IFxTransactionChannelType>(
      `${this.resourceUrl}/${getFxTransactionChannelTypeIdentifier(fxTransactionChannelType) as number}`,
      fxTransactionChannelType,
      { observe: 'response' }
    );
  }

  partialUpdate(fxTransactionChannelType: IFxTransactionChannelType): Observable<EntityResponseType> {
    return this.http.patch<IFxTransactionChannelType>(
      `${this.resourceUrl}/${getFxTransactionChannelTypeIdentifier(fxTransactionChannelType) as number}`,
      fxTransactionChannelType,
      { observe: 'response' }
    );
  }

  find(id: number): Observable<EntityResponseType> {
    return this.http.get<IFxTransactionChannelType>(`${this.resourceUrl}/${id}`, { observe: 'response' });
  }

  query(req?: any): Observable<EntityArrayResponseType> {
    const options = createRequestOption(req);
    return this.http.get<IFxTransactionChannelType[]>(this.resourceUrl, { params: options, observe: 'response' });
  }

  delete(id: number): Observable<HttpResponse<{}>> {
    return this.http.delete(`${this.resourceUrl}/${id}`, { observe: 'response' });
  }

  search(req: SearchWithPagination): Observable<EntityArrayResponseType> {
    const options = createRequestOption(req);
    return this.http.get<IFxTransactionChannelType[]>(this.resourceSearchUrl, { params: options, observe: 'response' });
  }

  addFxTransactionChannelTypeToCollectionIfMissing(
    fxTransactionChannelTypeCollection: IFxTransactionChannelType[],
    ...fxTransactionChannelTypesToCheck: (IFxTransactionChannelType | null | undefined)[]
  ): IFxTransactionChannelType[] {
    const fxTransactionChannelTypes: IFxTransactionChannelType[] = fxTransactionChannelTypesToCheck.filter(isPresent);
    if (fxTransactionChannelTypes.length > 0) {
      const fxTransactionChannelTypeCollectionIdentifiers = fxTransactionChannelTypeCollection.map(
        fxTransactionChannelTypeItem => getFxTransactionChannelTypeIdentifier(fxTransactionChannelTypeItem)!
      );
      const fxTransactionChannelTypesToAdd = fxTransactionChannelTypes.filter(fxTransactionChannelTypeItem => {
        const fxTransactionChannelTypeIdentifier = getFxTransactionChannelTypeIdentifier(fxTransactionChannelTypeItem);
        if (
          fxTransactionChannelTypeIdentifier == null ||
          fxTransactionChannelTypeCollectionIdentifiers.includes(fxTransactionChannelTypeIdentifier)
        ) {
          return false;
        }
        fxTransactionChannelTypeCollectionIdentifiers.push(fxTransactionChannelTypeIdentifier);
        return true;
      });
      return [...fxTransactionChannelTypesToAdd, ...fxTransactionChannelTypeCollection];
    }
    return fxTransactionChannelTypeCollection;
  }
}
